pkg/kubelet/cm/pod_container_manager_linux.go

```diff
@@ -108,10 +114,119 @@ func (m *podContainerManagerImpl) GetPodContainerName(pod *api.Pod) string {
+func (m *podContainerManagerImpl) getPIDsToKill(podCgroup string) []int {
+	// Get a list of processes that we need to kill
+	pidsToKill := sets.NewInt()
+	var pids []int
+	for _, val := range m.subsystems.MountPoints {
+		dir := path.Join(val, podCgroup)
+		_, err := os.Stat(dir)
+		if os.IsNotExist(err) {
+			// The subsystem pod cgroup is already deleted
+			// do nothing, continue
+			continue
+		}
+		// Get a list of pids that are still charged to the pod's cgroup
+		pids, err = readProcsFile(dir)
+		if err != nil {
+			continue
+		}
+		pidsToKill.Insert(pids...)
+
+		// WalkFunc which is called for each file and directory in the pod cgroup dir
+		visitor := func(path string, info os.FileInfo, err error) error {
+			if !info.IsDir() {
+				return nil
+			}
+			pids, err = readProcsFile(path)
+			if err != nil {
+				return err
+			}
+			pidsToKill.Insert(pids...)
+			return nil
+		}
+		// Walk through the pod cgroup directory to check if
+		// container cgroups haven't been GCed yet. Get attached processes to
+		// all such unwanted containers under the pod cgroup
+		err = filepath.Walk(dir, visitor)
+	}
+	return pidsToKill.List()
+}
+
+// Scan through the whole cgroup directory and kill all processes either
+// attached to the pod cgroup or to a container cgroup under the pod cgroup
+func (m *podContainerManagerImpl) tryKillingCgroupProcesses(podCgroup string) error {
+	pidsToKill := m.getPIDsToKill(podCgroup)
+	// No pids charged to the terminated pod cgroup return
+	if len(pidsToKill) == 0 {
+		return nil
+	}
+
+	var errlist []error
+	// os.Kill often errors out,
+	// We try killing all the pids multiple times
+	for i := 0; i < 5; i++ {
+		if i != 0 {
+			glog.V(3).Infof("Attempt %v failed to kill all unwanted process. Retyring", i)
+		}
+		errlist = []error{}
+		for _, pid := range pidsToKill {
+			p, err := os.FindProcess(pid)
+			if err != nil {
+				// Process not running anymore, do nothing
+				continue
+			}
+			glog.V(3).Infof("Attempt to kill process with pid: %v", pid)
+			if err := p.Kill(); err != nil {
+				glog.V(3).Infof("failed to kill process with pid: %v", pid)
+				errlist = append(errlist, err)
+			}
+		}
+		if len(errlist) == 0 {
+			glog.V(3).Infof("successfully killed all unwanted processes.")
+			return nil
+		}
+	}
+	return utilerrors.NewAggregate(errlist)
+}
+
 // Destroy destroys the pod container cgroup paths
 func (m *podContainerManagerImpl) Destroy(podCgroup string) error {
-	// This will house the logic for destroying the pod cgroups.
-	// Will be handled in the next PR.
+	// Set the pod's cgroup memory limit to a really small value.
+	// At this point the pod should completely free up cpu and memory resources.
+	// If there still exist tasks whose charges are being charged to the pod cgroup
+	// We will try killing them. Setting such a low memory limit on the pod cgroup
+	// might trigger OOMKills, but at this point that would be the desired behaviour
+	memoryLimit := minimumMemoryValue
```

Comment:
This is not possible. AFAIK, memcg will reject limits lower that usage. instead purge pages first and lower limit in steps, before deletion.